### ISSUE TITLE
ERROR: pgroonga: object isn't found: <Sources18956> on a Postres 16 DB

### ISSUE THREAD

**Issue by ghevge** [pgroonga/pgroonga#386]
### What happened?

I have a groonga/pgroonga:3.1.5-debian-16 DB instance which was created new, and populated with about 50GB worth of data. For a while everything was good but yesterday I started to notice some slowness in one of the queries so I ended up creating the index below:

`CREATE INDEX IF NOT EXISTS products_descriptions_product_id_idx ON products_descriptions (product_id);`

Right after creating that index I notices a crash  at the system level and the DB ended up in a weird state (I was able to connect to it using dbeaver but was not able to retrieve any tables related data and/or metadata. Nor was my java app able to fetch anything from it) from which it recovered only after restarting the system. Restarting just the containers didn't helped.

Now after I managed to restart it, I noticed that my java app was not happy anymore inserting values in the table were the index was created. The errors can be seen below

Googleing for the error I found https://github.com/pgroonga/pgroonga/issues/335 so I've tried to enable pgroonga.enable_wal but after restarting the DB, I still see the same errors.

I tried to remove and recreate the new index, while the wal was enabled, but was still seeing the same behavior.
Even if I remove the new index altogether, I still get the same error.  

Any idea how I can get rid of this error?

Thanks

My postgres.conf pgroonga related configs:
> shared_preload_libraries = 'pgroonga_crash_safer'
> pgroonga.enable_crash_safe = on
> pgroonga.enable_wal = on

products_description table DDL:
```
CREATE TABLE public.products_descriptions (
	id uuid NOT NULL,
	product_id uuid NOT NULL,
	locale varchar(20) NOT NULL,
	title varchar(500) NOT NULL,
	description varchar(11000) NULL,
	username varchar(255) NOT NULL,
	creation_time timestamp NOT NULL,
	update_time timestamp NOT NULL,
	short_description varchar(450) NULL,
	commit_id uuid NULL,
	CONSTRAINT product_description_pkey PRIMARY KEY (id),
	CONSTRAINT products_descriptions_product_id_locale_key UNIQUE (product_id, locale),
	CONSTRAINT products_descriptions_product_id_fkey FOREIGN KEY (product_id) REFERENCES public.products(id) ON DELETE CASCADE
);
CREATE INDEX product_description_title_idx ON public.products_descriptions USING pgroonga (title);
CREATE INDEX products_descriptions_id_idx ON public.products_descriptions USING btree (id);
CREATE INDEX products_descriptions_locale_idx ON public.products_descriptions USING btree (locale);
CREATE INDEX products_descriptions_product_id_idx ON public.products_descriptions USING btree (product_id);
```

Observed error:

```
2023-12-24T15:22:51.830Z <> {thread=pool-7-thread-7} ERROR: [com.kp.mw.services.TranslationsService] com.kp.mw.services.TranslationsServicerlf(Failed to translate product) 
org.springframework.dao.DataIntegrityViolationException: 
### Error updating database.  Cause: org.postgresql.util.PSQLException: ERROR: pgroonga: object isn't found: <Sources18956>
### The error may exist in com/kp/mw/db/mappers/ProductsDescriptionsMapper.xml
### The error may involve com.kp.mw.db.mappers.ProductsDescriptionsMapper.createProductDescription-Inline
### The error occurred while setting parameters
### SQL: INSERT INTO products_descriptions (id, username, creation_time, update_time, product_id, locale, title, description, short_description, commit_id)         VALUES (?, ?, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP, ?, ?, ?, ?, ?, ?)         ON CONFLICT (product_id, locale) DO UPDATE SET username =  ?, title = ?, description = ?, short_description = ?, update_time = CURRENT_TIMESTAMP
### Cause: org.postgresql.util.PSQLException: ERROR: pgroonga: object isn't found: <Sources18956>
; ERROR: pgroonga: object isn't found: <Sources18956>
	at org.springframework.jdbc.support.SQLStateSQLExceptionTranslator.doTranslate(SQLStateSQLExceptionTranslator.java:118) ~[spring-jdbc-6.1.1.jar!/:6.1.1]
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:107) ~[spring-jdbc-6.1.1.jar!/:6.1.1]
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:116) ~[spring-jdbc-6.1.1.jar!/:6.1.1]
	at org.springframework.jdbc.support.AbstractFallbackSQLExceptionTranslator.translate(AbstractFallbackSQLExceptionTranslator.java:116) ~[spring-jdbc-6.1.1.jar!/:6.1.1]
	at org.mybatis.spring.MyBatisExceptionTranslator.translateExceptionIfPossible(MyBatisExceptionTranslator.java:92) ~[mybatis-spring-3.0.3.jar!/:3.0.3]
	at org.mybatis.spring.SqlSessionTemplate$SqlSessionInterceptor.invoke(SqlSessionTemplate.java:439) ~[mybatis-spring-3.0.3.jar!/:3.0.3]
	at jdk.proxy2/jdk.proxy2.$Proxy120.insert(Unknown Source) ~[?:?]
	at org.mybatis.spring.SqlSessionTemplate.insert(SqlSessionTemplate.java:272) ~[mybatis-spring-3.0.3.jar!/:3.0.3]
	at org.apache.ibatis.binding.MapperMethod.execute(MapperMethod.java:62) ~[mybatis-3.5.14.jar!/:3.5.14]
	at org.apache.ibatis.binding.MapperProxy$PlainMethodInvoker.invoke(MapperProxy.java:141) ~[mybatis-3.5.14.jar!/:3.5.14]
	at org.apache.ibatis.binding.MapperProxy.invoke(MapperProxy.java:86) ~[mybatis-3.5.14.jar!/:3.5.14]
	at jdk.proxy2/jdk.proxy2.$Proxy157.createProductDescription(Unknown Source) ~[?:?]
	at com.kp.mw.services.ProductsDescriptionsService.createProductDescription(ProductsDescriptionsService.java:33) ~[!/:?]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:352) ~[spring-aop-6.1.1.jar!/:6.1.1]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196) ~[spring-aop-6.1.1.jar!/:6.1.1]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163) ~[spring-aop-6.1.1.jar!/:6.1.1]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:765) ~[spring-aop-6.1.1.jar!/:6.1.1]
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:123) ~[spring-tx-6.1.1.jar!/:6.1.1]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:385) ~[spring-tx-6.1.1.jar!/:6.1.1]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:119) ~[spring-tx-6.1.1.jar!/:6.1.1]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184) ~[spring-aop-6.1.1.jar!/:6.1.1]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:765) ~[spring-aop-6.1.1.jar!/:6.1.1]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:717) ~[spring-aop-6.1.1.jar!/:6.1.1]
	at com.kp.mw.services.ProductsDescriptionsService$$SpringCGLIB$$0.createProductDescription(<generated>) ~[!/:?]
	at com.kp.mw.services.TranslationsService.translateProduct(TranslationsService.java:954) ~[!/:?]
	at com.kp.mw.services.TranslationsService$ProcessTranslateProductRequest.run(TranslationsService.java:1012) ~[!/:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: org.postgresql.util.PSQLException: ERROR: pgroonga: object isn't found: <Sources18956>
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2676) ~[postgresql-42.5.4.jar!/:42.5.4]
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2366) ~[postgresql-42.5.4.jar!/:42.5.4]
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:356) ~[postgresql-42.5.4.jar!/:42.5.4]
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:496) ~[postgresql-42.5.4.jar!/:42.5.4]
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:413) ~[postgresql-42.5.4.jar!/:42.5.4]
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:190) ~[postgresql-42.5.4.jar!/:42.5.4]
	at org.postgresql.jdbc.PgPreparedStatement.execute(PgPreparedStatement.java:177) ~[postgresql-42.5.4.jar!/:42.5.4]
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.execute(ProxyPreparedStatement.java:44) ~[HikariCP-5.1.0.jar!/:?]
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.execute(HikariProxyPreparedStatement.java) ~[HikariCP-5.1.0.jar!/:?]
	at io.opentelemetry.instrumentation.jdbc.internal.OpenTelemetryStatement.wrapCall(OpenTelemetryStatement.java:294) ~[opentelemetry-jdbc-1.29.0-alpha.jar!/:1.29.0-alpha]
	at io.opentelemetry.instrumentation.jdbc.internal.OpenTelemetryPreparedStatement.execute(OpenTelemetryPreparedStatement.java:70) ~[opentelemetry-jdbc-1.29.0-alpha.jar!/:1.29.0-alpha]
	at org.apache.ibatis.executor.statement.PreparedStatementHandler.update(PreparedStatementHandler.java:48) ~[mybatis-3.5.14.jar!/:3.5.14]
	at org.apache.ibatis.executor.statement.RoutingStatementHandler.update(RoutingStatementHandler.java:75) ~[mybatis-3.5.14.jar!/:3.5.14]
	at org.apache.ibatis.executor.SimpleExecutor.doUpdate(SimpleExecutor.java:50) ~[mybatis-3.5.14.jar!/:3.5.14]
	at org.apache.ibatis.executor.BaseExecutor.update(BaseExecutor.java:117) ~[mybatis-3.5.14.jar!/:3.5.14]
	at org.apache.ibatis.executor.CachingExecutor.update(CachingExecutor.java:76) ~[mybatis-3.5.14.jar!/:3.5.14]
	at org.apache.ibatis.session.defaults.DefaultSqlSession.update(DefaultSqlSession.java:197) ~[mybatis-3.5.14.jar!/:3.5.14]
	at org.apache.ibatis.session.defaults.DefaultSqlSession.insert(DefaultSqlSession.java:184) ~[mybatis-3.5.14.jar!/:3.5.14]
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at org.mybatis.spring.SqlSessionTemplate$SqlSessionInterceptor.invoke(SqlSessionTemplate.java:425) ~[mybatis-spring-3.0.3.jar!/:3.0.3]
	... 25 more
```

### How to reproduce it

See above. Unfortunately I can't share the whole DB schema.

### Expected behavior

No error

### Environment

groonga/pgroonga:3.1.5-debian-16

**Comment by kou** [comment]
Could you provide a schema for `products`?
Or could you check whether this is happen without `CONSTRAINT products_descriptions_product_id_fkey FOREIGN KEY (product_id) REFERENCES products(id) ON DELETE CASCADE` or not?

**Comment by kou** [comment]
The followings may recover this situation:

1. Stop PostgreSQL
2. Remove all `pgrn*` files in PostgreSQL's data directory
3. Start PostgreSQL
4. `REINDEX DATABASE YOUR_DATABASE` (or reindex only PGroonga indexes)

**Comment by kou** [comment]
Could you also provide full `pgroonga.log`?

**Comment by ghevge** [comment]
I can share my pgroonga.log the only problem is that it only has about 12GB. Do you have an FTP or another location that can receive such big files? 

As for the products schema, I'm pasting in below. I will try removing the foreign key to see if that helps in any way.

My problem with you proposed database reindexing is that, it will work while testing the app, but if this will occur in production, I won't be able to afford that. Plus the DB is not up to its targeted size yet. The prod DB will easily have a couple of hundreds GBs if not TBs of data. Reindexing all that data won't be fast. 

Thanks,
Alin


CREATE TABLE public.products (
	id uuid NOT NULL,
	mass numeric NULL,
	display_mass_unit varchar(10) NULL,
	length numeric NULL,
	width numeric NULL,
	height numeric NULL,
	display_lenght_unit varchar(10) NULL,
	volume numeric NULL,
	display_volume_unit varchar(10) NULL,
	category int2 NOT NULL,
	creation_time timestamp NOT NULL,
	update_time timestamp NOT NULL,
	username varchar(255) NOT NULL,
	nutrition_facts_per_quantity numeric NULL,
	nutrition_facts_unit varchar(10) NULL DEFAULT NULL::character varying,
	rating numeric NULL,
	normalized_length numeric NULL,
	normalized_width numeric NULL,
	normalized_height numeric NULL,
	normalized_volume numeric NULL,
	normalized_mass numeric NULL,
	link1 varchar(2048) NULL DEFAULT NULL::character varying,
	link2 varchar(2048) NULL DEFAULT NULL::character varying,
	link3 varchar(2048) NULL DEFAULT NULL::character varying,
	discontinued bool NULL DEFAULT false,
	has_images bool NULL DEFAULT false,
	health_score numeric NULL,
	efficiency_score numeric NULL,
	eco_score numeric NULL,
	commit_id uuid NULL,
	production_cost numeric NULL,
	video1 varchar(2048) NULL DEFAULT NULL::character varying,
	video2 varchar(2048) NULL DEFAULT NULL::character varying,
	video3 varchar(2048) NULL DEFAULT NULL::character varying,
	CONSTRAINT product_pkey PRIMARY KEY (id)
);
CREATE INDEX products_category_idx ON public.products USING btree (category);
CREATE INDEX products_id_idx ON public.products USING btree (id);

**Comment by ghevge** [comment]
I've remove the CONSTRAINT products_descriptions_product_id_fkey  but I'm still seeing the same behavior

**Comment by ghevge** [comment]
I've removed all the pgroonga indexes from the products_descriptions table and the error went away. I'm trying to recreate now those pgroonga indexes on that table, but as expected this operation is slow... Won't fly in production. I wonder if there is a way to somehow have a failover mechanism in pgroonga to promote new indexes only after they are validated to be ok. And keep the older stable ones, until the validation for the new ones is completed.

**Comment by askdkc** [comment]
@ghevge 

PostgreSQL can create as many streaming replications as you want. And PGroonga is also supported. So you might try using one of the replications for production and others for backup.

https://pgroonga.github.io/reference/streaming-replication.html

Or you can create a new db.
Then dump and restore all the data to newly created db.
Create index.
Switch production db to the new one. 
But if your db is constantly getting inserted many data, this approach is not suitable.

**Comment by ghevge** [comment]
The DB will get constantly updated (inserts, updates, deletes)

**Comment by ghevge** [comment]
from what I can see, all the pgroonga indexes were affected. Once I fixed the products_descriptions table other similar errors started to pop on other tables containing pgroonga indexes.

**Comment by ghevge** [comment]
I was looking under the postgres data folder but I  am not seeing any pgrn* files in there

> [root@localhost db]# ls
> base    pg_commit_ts  pg_hba.conf    pg_logical    pg_notify    pgroonga.log  pg_snapshots  pg_stat_tmp  pg_tblspc    PG_VERSION  pg_xact               postgresql.conf  postmaster.pid
> global  pg_dynshmem   pg_ident.conf  pg_multixact  pg_replslot  pg_serial     pg_stat       pg_subtrans  pg_twophase  pg_wal      postgresql.auto.conf  postmaster.opts
> [root@localhost db]# ls | grep pgrn*
> pgroonga.log
> [root@localhost db]#

I found some though under 'base' folder:

> [root@localhost db]# find . -name "pgrn*"
> ./base/16384/pgrn.000010C
> ./base/16384/pgrn.0000102
> ./base/16384/pgrn.0000120.c
> ./base/16384/pgrn.0000115
> ./base/16384/pgrn.0000126
> ./base/16384/pgrn.000012A
> ./base/16384/pgrn.options
> ./base/16384/pgrn.0000124
> ./base/16384/pgrn.0000128
> ./base/16384/pgrn.0000127
> ./base/16384/pgrn.0000129
> ./base/16384/pgrn.0000113
> ./base/16384/pgrn.0000125
> ./base/16384/pgrn.000010D
> ./base/16384/pgrn.0000120
> ./base/16384/pgrn.conf
> ./base/16384/pgrn.0000121
> ./base/16384/pgrn.000011F
> ./base/16384/pgrn.000011D
> ./base/16384/pgrn.0000108
> ./base/16384/pgrn.0000105
> ./base/16384/pgrn.000010E.c
> ./base/16384/pgrn.000010B
> ./base/16384/pgrn.0000107
> ./base/16384/pgrn.000010E
> ./base/16384/pgrn.000010F
> ./base/16384/pgrn.0000124.c
> ./base/16384/pgrn.0000103
> ./base/16384/pgrn.0000110
> ./base/16384/pgrn.000011F.c
> ./base/16384/pgrn.000010A
> ./base/16384/pgrn.0000109
> ./base/16384/pgrn.0000116
> ./base/16384/pgrn.0000000
> ./base/16384/pgrn.000012B
> ./base/16384/pgrn.000012C
> ./base/16384/pgrn.0000114
> ./base/16384/pgrn.0000122
> ./base/16384/pgrn.000010A.c
> ./base/16384/pgrn.000012C.c
> ./base/16384/pgrn.0000104
> ./base/16384/pgrn.0000123
> ./base/16384/pgrn.001
> ./base/16384/pgrn.0000101
> ./base/16384/pgrn.0000112.c
> ./base/16384/pgrn.0000128.c
> ./base/16384/pgrn.0000112
> ./base/16384/pgrn.0000111
> ./base/16384/pgrn.000011E
> ./base/16384/pgrn
> [root@localhost db]#

**Comment by kou** [comment]
> Are these the files you were asking to remove before reindexing the DB ?

Yes.

**Comment by kou** [comment]
> I wonder if there is a way to somehow have a failover mechanism in pgroonga to promote new indexes only after they are validated to be ok. And keep the older stable ones, until the validation for the new ones is completed.

You can use `REINDEX CONCURRENTLY` for it: https://www.postgresql.org/docs/current/sql-reindex.html

**Comment by kou** [comment]
> I can share my pgroonga.log the only problem is that it only has about 12GB. Do you have an FTP or another location that can receive such big files?

Could you try `zstd pgroonga.log` and check the compressed size? Can we attach the compressed log here?

**Comment by ghevge** [comment]
> > I can share my pgroonga.log the only problem is that it only has about 12GB. Do you have an FTP or another location that can receive such big files?
> 
> Could you try `zstd pgroonga.log` and check the compressed size? Can we attach the compressed log here?

The zstd compressed file is 1.3 GB. Github only allows 25 MB to be uploaded

**Comment by kou** [comment]
OK. Can we use https://gist.github.com/ for the compressed file?

**Comment by ghevge** [comment]
> OK. Can we use https://gist.github.com/ for the compressed file?

From what I can see this is more tailored fore notes and code snippet. I don't think this is the right tool to share such big files. Do you have a gmail account? I can share it through my google drive

**Comment by kou** [comment]
OK. Could you use koutou@gmail.com for it?

**Comment by ghevge** [comment]
Sent it. Let me know if you got it.

**Comment by ghevge** [comment]
FYI I've used zsdt for windows to pack it : https://sourceforge.net/projects/zstd-for-windows/

**Comment by ghevge** [comment]
BTW a question slightly outside of this topi: Any particular reason why the log file is allowed to grow so big? Won't be better to split / roll it ?

**Comment by kou** [comment]
Thanks. I could receive it and uncompressed it.

**Comment by ghevge** [comment]
> Thanks. I could receive it and uncompressed it.

Super! Let me know if you need anything else. The DB is still partially in this weird state as I haven't removed the prgn files yet. I will probably do it tomorrow . So if you need anything else before I will proceed with the cleanup, let me know.

**Comment by kou** [comment]
> BTW a question slightly outside of this topi: Any particular reason why the log file is allowed to grow so big? Won't be better to split / roll it ?

It's intended to use external log rotation tool such as logrotate.

Or we can use PostgreSQL's logging feature instead of logging to `pgroonga.log`: https://pgroonga.github.io/reference/parameters/log-type.html

Or we can add a configuration to enable log rotation feature provided by Groonga. If you want it, could you open a new issue for it.

**Comment by ghevge** [comment]
> If you want it, could you open a new issue for it.

Is not about me wanting it. Usually the systems get slower the bigger the log files get plus the is the space problem. With this behavior you will get out of space quicker. I would say that it will be in the benefit of everyone's using pgroonga to have a log rotation/cleanup mechanism by default on a pgroonga instance. If it wouldn't have been for this issue, I wouldn't have realized logs related problem.  Anyhow I would open another issue for this, tomorrow.

**Comment by kou** [comment]
Our deb packages provide log rotate configuration: https://github.com/pgroonga/pgroonga/blob/main/packages/postgresql-16-pgdg-pgroonga/debian/postgresql-16-pgdg-pgroonga.logrotate

But logrotate isn't used in a Docker container because PostgreSQL processes are only running in a Docker container.
If it's a Docker image only problem, https://github.com/pgroonga/docker/issues is suitable than https://github.com/pgroonga/pgroonga/issue .

**Comment by ghevge** [comment]
@kou something is wrong with the pgroonga indexes! After reindexing my db (which took 3 hours to complete), I started a new data injection session, which run for about 20 hours and inserted a couple of million new records. All good there until I decided to restart the DB container, in order to increase the max number of allowed connections so that i can start new application nodes to inject data. 

So I stopped the DB instance using: docker stop <db id> , and then restarted it using docker-compose. There was no error thrown during stop or restart operations. 

The DB instance did restart, and tables which don't have  pgroonga indexes define are accessible, but the table with pgroonga indexes, are not accessible. From dbeaver side I can see errors like :

```
org.jkiss.dbeaver.model.sql.DBSQLException: SQL Error [58000]: ERROR: pgroonga: already tried to initialize and failed
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:133)
	at org.jkiss.dbeaver.model.impl.jdbc.struct.JDBCTable.readData(JDBCTable.java:187)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.lambda$0(ResultSetJobDataRead.java:123)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:189)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.run(ResultSetJobDataRead.java:121)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetViewer$ResultSetDataPumpJob.run(ResultSetViewer.java:5101)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:105)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: org.postgresql.util.PSQLException: ERROR: pgroonga: already tried to initialize and failed
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2676)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2366)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:356)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:496)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:413)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:333)
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:319)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:295)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:290)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.execute(JDBCStatementImpl.java:330)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:131)
	... 7 more
```

I will send you over google drive the new set of pgroonga logs.

**Comment by ghevge** [comment]
Latest pgroonga logs sent over google drive!

**Comment by kou** [comment]
Thanks.

Workaround: Disable `pgroonga_crash_safer`:

```text
shared_preload_libraries = ''
pgroonga.enable_crash_safe = off
```

The "pgroonga: already tried to initialize and failed" error will be disappeared eventually. You don't need to do anything.

The error is happen because auto crash safe feature is still running. (Broken indexes are recreating.) 

The real problem is that `pgroonga_crash_safer` detects some indexes are broken.
In general, the background worker for `pgroonga_crash_safer` (PID: `43`) is finished after all other workers that process client connections. But one process (PID: `2029936`) finished after the background worker `pgroonga_crash_safer`. The `2029936` process might change some indexes and `pgroonga_crash_safer` marks them as broken indexes.

I'll add more logs to `pgroonga_crash_safer` and ensure `pgroonga_crash_safer` waits other connections are finished. Could you try them after I've done them?

**Comment by ghevge** [comment]
I will try to disable pgroonga_crash_safer and see. But the reason i have it enable is because you suggested it , a while ago while having some other issues I believe also related to indexes. I don't recall the exact problem now.

And yes, I can test whatever you want me to try. Just let me know. 

Also note I am on EST time. I guess you are on Japan time. So there is a 10 hour difference :)  when is morning there is night here and vice versa.

**Comment by ghevge** [comment]
OK I've disabled pgroonga_crash_safer and the DB is accessible again!

**Comment by kou** [comment]
> Also note I am on EST time. I guess you are on Japan time. So there is a 10 hour difference :)

Right. :-)

BTW, are you using streaming replication? If you don't use it, you don't need `pgroonga.enable_wal = on`.

**Comment by ghevge** [comment]
No I'm not using streaming replication. I enabled it before I've got your initial feedback, and forgot to turn it off.

**Comment by kou** [comment]
I think that we can fix this by 1fe626208fb50f2cd009a774b9564dbbfd8cf1e7.
We'll release a new version in a few weeks. I'll notify it to you when we does it.

**Comment by ghevge** [comment]
Thanks!

**Comment by kou** [comment]
Thanks.

Your workaround is correct because PGroonga doesn't enable WAL by default.

Using WAL resource manager may be alternative solution because it enables WAL.
See also: https://pgroonga.github.io/reference/modules/pgroonga-wal-resource-manager.html
But it has performance overhead for write. So it may not be suitable for testing.
